src/LinAlg/hiopVectorCuda.cpp

```diff
@@ -0,0 +1,1118 @@
+}
+
+/** @brief this[i] += alpha*x[i]*z[i] forall i */
+void hiopVectorCuda::axzpy(double alpha, const hiopVector& xvec, const hiopVector& zvec)
+{
+  const hiopVectorCuda& x = dynamic_cast<const hiopVectorCuda&>(xvec);
+  const hiopVectorCuda& z = dynamic_cast<const hiopVectorCuda&>(zvec);
+#ifdef HIOP_DEEPCHECKS
+  assert(x.n_local_ == z.n_local_);
+  assert(  n_local_ == z.n_local_);
+#endif  
+  double* dd       = data_dev_;
+  const double* xd = x.local_data_const();
+  const double* zd = z.local_data_const();
+
+  hiop::cuda::axzpy_kernel(n_local_, dd, xd, zd, alpha);
+}
+
+/** @brief this[i] += alpha*x[i]/z[i] forall i */
+void hiopVectorCuda::axdzpy(double alpha, const hiopVector& xvec, const hiopVector& zvec)
+{
+  const hiopVectorCuda& x = dynamic_cast<const hiopVectorCuda&>(xvec);
+  const hiopVectorCuda& z = dynamic_cast<const hiopVectorCuda&>(zvec);
+#ifdef HIOP_DEEPCHECKS
+  assert(x.n_local_ == z.n_local_);
+  assert(  n_local_ == z.n_local_);
+#endif  
+  double*       yd = data_dev_;
+  const double* xd = x.local_data_const();
+  const double* zd = z.local_data_const();
+
+  hiop::cuda::axdzpy_kernel(n_local_, yd, xd, zd, alpha);
+}
+
+/** @brief this[i] += alpha*x[i]/z[i] forall i with pattern selection */
+void hiopVectorCuda::axdzpy_w_pattern(double alpha,
+                                      const hiopVector& xvec,
+                                      const hiopVector& zvec,
+                                      const hiopVector& select)
+{
+  const hiopVectorCuda& x = dynamic_cast<const hiopVectorCuda&>(xvec);
+  const hiopVectorCuda& z = dynamic_cast<const hiopVectorCuda&>(zvec);
+  const hiopVectorCuda& sel = dynamic_cast<const hiopVectorCuda&>(select);
+#ifdef HIOP_DEEPCHECKS
+  assert(x.n_local_==z.n_local_);
+  assert(  n_local_==z.n_local_);
+#endif
+  double* yd = data_dev_;
+  const double* xd = x.local_data_const();
+  const double* zd = z.local_data_const();
+  const double* id = sel.local_data_const();
+
+  hiop::cuda::axdzpy_w_pattern_kernel(n_local_, yd, xd, zd, id, alpha);
+}
+
+/** @brief this[i] += c forall i */
+void hiopVectorCuda::addConstant(double c)
+{
+  hiop::cuda::add_constant_kernel(n_local_, data_dev_, c);
+}
+
+/** @brief this[i] += c forall i with pattern selection */
+void  hiopVectorCuda::addConstant_w_patternSelect(double c, const hiopVector& select)
+{
+  const hiopVectorCuda& sel = dynamic_cast<const hiopVectorCuda&>(select);
+  assert(this->n_local_ == sel.n_local_);
+  const double* id = sel.local_data_const();
+
+  hiop::cuda::add_constant_w_pattern_kernel(n_local_, data_dev_, id, c);
+}
+
+/** Return the dot product of this hiopVector with v */
+double hiopVectorCuda::dotProductWith( const hiopVector& v ) const
+{
+  const hiopVectorCuda& vx = dynamic_cast<const hiopVectorCuda&>(v);
+  int one = 1;
+  double retval; 
+  cublasStatus_t ret_cublas = cublasDdot(handle_cublas_, n_local_, vx.data_dev_, one, data_dev_, one, &retval);
+  assert(ret_cublas == CUBLAS_STATUS_SUCCESS);
```

Comment:
If you already use Thrust, why use cuBLAS here?

Reply:
My strategy is to use Thrust if no cuBlas function is available. If neither is available, then implement our own kernel functions.